### PR TITLE
scrcpy 1.1 (new formula)

### PR DESCRIPTION
**Pull request by stek29**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

**Comment by stek29** [comment]
scrcpy has a "server" which is pushed to device and ran on it.
However, building it requires android sdk, which is only available as cask, is heavy and requires additional configuration and user interaction before usage.

So instead I opted out for just using prebuilt server provided by maintainer.

**Comment by AdoPi** [comment]
Haha, I was expecting to put the same formula, nice job :)

File: Formula/scrcpy.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+  resource "prebuilt-server" do
+    url "https://github.com/Genymobile/scrcpy/releases/download/v1.0/scrcpy-server-v1.0.jar"
+    sha256 "b573b06a6072476b85b6308e3ad189f2665ad5be4f8ca3a6b9ec81d64df20558"
+  end
+
+  def install
+    r = resource("prebuilt-server")
+    r.verify_download_integrity(r.fetch)
+    cp r.fetch, buildpath/"prebuilt-server.jar"
```
Comment (fxcoudert) [inline Formula/scrcpy.rb]: `resource("prebuilt-server").stage buildpath/"prebuilt-server.jar"` should work and is simpler

File: Formula/scrcpy.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    meson_args = %w[x --buildtype release --strip -Db_lto=true]
+    meson_args << "-Dprebuilt_server=#{buildpath/"prebuilt-server.jar"}"
+    meson_args << "--prefix=#{prefix}"
+
+    system "meson", *meson_args
```
Comment (fxcoudert) [inline Formula/scrcpy.rb]: No extra variable, just inline the args in the system call

File: Formula/scrcpy.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    url "https://github.com/Genymobile/scrcpy/releases/download/v1.0/scrcpy-server-v1.0.jar"
+    sha256 "b573b06a6072476b85b6308e3ad189f2665ad5be4f8ca3a6b9ec81d64df20558"
+  end
+
+  def install
+    r = resource("prebuilt-server")
+    r.verify_download_integrity(r.fetch)
+    cp r.fetch, buildpath/"prebuilt-server.jar"
+
+    meson_args = %w[x --buildtype release --strip -Db_lto=true]
```
Comment (fxcoudert) [inline Formula/scrcpy.rb]: What does this lone `x` mean?
If it's an arbitrary build directory, can be use `build` instead please?

File: Formula/scrcpy.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    meson_args << "-Dprebuilt_server=#{buildpath/"prebuilt-server.jar"}"
+    meson_args << "--prefix=#{prefix}"
+
+    system "meson", *meson_args
+
+    chdir "x" do
```
Comment (fxcoudert) [inline Formula/scrcpy.rb]: `cd` instead of `chdir`

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/scrcpy.rb (hunk 1 above)
Comment (stek29) [reply Formula/scrcpy.rb]: It would save file to `buildpath/"prebuilt-server.jar"/"scrcpy-server-v1.0.jar"`, not to `buildpath/"prebuilt-server.jar"`

Comment (stek29) [reply Formula/scrcpy.rb]: I also tried to do something like this:
```
resource "prebuilt-server" do
    url "https://github.com/Genymobile/scrcpy/releases/download/v#{version}/scrcpy-server-v#{version}.jar"
    sha256 "b573b06a6072476b85b6308e3ad189f2665ad5be4f8ca3a6b9ec81d64df20558"
end
```

But obviously version was nil inside of block, so it won't work :\

**Comment by stek29** [comment]
@fxcoudert ping

**Comment by nvllsvm** [comment]
@stek29 - Version 1.1 is out now and includes a fixes a rather gruesome memory leak. Might as well change this to that version.

**Comment by stek29** [comment]
@fxcoudert .fetch does unpack, which means that I'd have to duplicate filename or do some magical stuff on resource url, which definitely won't be cleaner than current solution.

**Comment by stek29** [comment]
Hello?..

File: Formula/scrcpy.rb (hunk 1 above)
Comment (fxcoudert) [reply Formula/scrcpy.rb]: I think we want `cp r.cached_download` instead of `cp r.fetch`

File: Formula/scrcpy.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    sha256 "14826512bf38447ec94adf3b531676ce038d19e7e06757fb4e537882b17e77b3"
+  end
+
+  def install
+    r = resource("prebuilt-server")
+    r.verify_download_integrity(r.fetch)
+    cp r.fetch, buildpath/"prebuilt-server.jar"
+
+    system "meson", "build", "--buildtype", "release",
+      "--strip", "-Db_lto=true", "--prefix=#{prefix}",
```
Comment (fxcoudert) [inline Formula/scrcpy.rb]: Most of our other meson-based formulas do:

```
mkdir "build" do
  system "meson", "--prefix=...", ".."
  system "ninja", "install"
end
```

So I'm wondering:
- why are we passing `build` here?
- if `--strip` and `-Db_lto=true` are not default, why use them?
- is `--buildtype release` necessary or is it the default?

(And in any case, the indentation for continuation lines is wrong)

**Comment by stek29** [comment]
@fxcoudert 
> And in any case, the indentation for continuation lines is wrong

Hm, is it still wrong? What is correct indentation? `brew audit` doesn't complain...

**Comment by fxcoudert** [comment]
```
    system "meson", "..", "--prefix=#{prefix}",
                             "-Dprebuilt_server=#{buildpath/"prebuilt-server.jar"}"
```

File: Formula/scrcpy.rb (hunk 5 above)
Comment (rom1v) [reply Formula/scrcpy.rb]: > if `--strip` and `-Db_lto=true` are not default, why use them?

They are not absolutely necessary, but why not enabling link-time optimization (even if meson does not enable it by default)?

> is `--buildtype release` necessary or is it the default?

The [default](http://mesonbuild.com/Running-Meson.html#configuring-the-source) is debug.

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thanks @stek29!

**Comment by rom1v** [comment]
Thank you very much :heart: That's great :+1: 

However, the package installed via homebrew is in debug mode (it prints debug logs and have the debug icon) instead of release mode (`--buildtype release` parameter is missing as meson arguments).

**Comment by rom1v** [comment]
@fxcoudert @stek29 Thank you very much :heart: That's great :+1: 

However, the package installed via homebrew is in debug mode (it prints debug logs and have the debug icon) instead of release mode (`--buildtype release` parameter is missing as meson arguments).

**Comment by fxcoudert** [comment]
@rom1v it seems to be the default mode, so maybe you can ask upstream to switch it?

**Comment by rom1v** [comment]
@fxcoudert This is the default in meson. AFAIK there is no way to switch the default (it's up to the caller), or maybe I missed something.

(I'm the upstream :wink:)

**Pull request closed without merging**